### PR TITLE
Sparse matrix python interface

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**
this adds the python interface for the serial Kratos CSR and system vector (mpi implementation will follow in a different PR)

Please mark the PR with appropriate tags: 
- improvement

**🆕 Changelog**
python interface for the serial Kratos CSR and system vector (mpi implementation will follow in a different PR)

### CODE REVIEW

File: kratos/containers/system_vector.h [hunk 1]
```diff
@@ -144,7 +144,17 @@ class SystemVector final
         return mData[I];
     }
 
-    void Add(const double factor,
+    DenseVector<TDataType>& data()
```
Comment (loumalouomega) [inline kratos/containers/system_vector.h:148]: Some definitions?

File: kratos/python/add_sparse_matrices_to_python.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+// Project includes
+#include "includes/define_python.h"
+#include "includes/ublas_interface.h"
+#include "python/add_sparse_matrices_to_python.h"
+#include "containers/sparse_graph.h"
+#include "containers/sparse_contiguous_row_graph.h"
+#include "containers/csr_matrix.h"
+#include "containers/system_vector.h"
+#include "utilities/amgcl_csr_spmm_utilities.h"
+namespace Kratos
```
Comment (loumalouomega) [inline kratos/python/add_sparse_matrices_to_python.cpp:28]:
```suggestion

namespace Kratos
```
Missing space

File: kratos/python/add_sparse_matrices_to_python.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+       self.Assemble(values,indices);
+    })
+
+    //inplace
+    .def("__iadd__", [](SystemVector<double,IndexType>& self, const SystemVector<double,IndexType>& other_vec){self += other_vec; }, py::is_operator())
+    .def("__isub__", [](SystemVector<double,IndexType>& self, const SystemVector<double,IndexType>& other_vec){self -= other_vec;  }, py::is_operator())
+    .def("__imul__", [](SystemVector<double,IndexType>& self, const double& value){ self*=value; }, py::is_operator())
+    .def("__itruediv__", [](SystemVector<double,IndexType>& self, const double& value){ self/=value; }, py::is_operator())
+
+    //out of place
```
Comment (loumalouomega) [inline kratos/python/add_sparse_matrices_to_python.cpp:201]: Clean unused?

File: kratos/python/add_sparse_matrices_to_python.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Pooyan Dadvand
+//                   Riccardo Rossi
+//
+
+
```
Comment (loumalouomega) [inline kratos/python/add_sparse_matrices_to_python.h:14]: Remove extra \n?

File: kratos/tests/test_sparse_matrices.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
+import KratosMultiphysics
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+import KratosMultiphysics.kratos_utilities as kratos_utils
+import numpy as np
```
Comment (loumalouomega) [inline kratos/tests/test_sparse_matrices.py:5]:
```suggestion

import numpy as np 
```

Separate Kratos imports of other imports

File: kratos/tests/test_sparse_matrices.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        self.assertEqual(y[2],6.0)
+
+        y = KratosMultiphysics.SystemVector(3)
+        y.SetValue(0.0)
+        A.TransposeSpMV(x,y)
+        self.assertEqual(y[0],5.0)
+        self.assertEqual(y[1],0.0)
+        self.assertEqual(y[2],13.0)
+
+
```
Comment (loumalouomega) [inline kratos/tests/test_sparse_matrices.py]: Extra \n

File: kratos/tests/test_sparse_matrices.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        B = A.SpMM(A)
+
+        for i in range(len(validation_data)):
+            self.assertEqual(B.value_data()[i], validation_data[i])
+            self.assertEqual(B.index2_data()[i], validation_index2[i])
+
+        for i in range(len(validation_index1)):
+            self.assertEqual(B.index1_data()[i], validation_index1[i])
+
+        # the following should be added back in case scipy support is enabled in testing
```
Comment (loumalouomega) [inline kratos/tests/test_sparse_matrices.py:89]: I would add a test that is skipped if no scipy, instead of a commented block

File: kratos/tests/test_sparse_matrices.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        # for i in range(len(validation_data)):
+        #     self.assertEqual(B_scipy.data[i], validation_data[i])
+        #     self.assertEqual(B_scipy.index2[i], validation_index2[i])
+
+        # for i in range(len(validation_index1)):
+        #     self.assertEqual(B_scipy.indptr[i], validation_index1[i])
+        # except ImportError:
+        #     pass
+       
+
```
Comment (loumalouomega) [inline kratos/tests/test_sparse_matrices.py:108]: Extra \n

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Somme coments

File: kratos/tests/test_sparse_matrices.py (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/tests/test_sparse_matrices.py:89]: no, i prefer this way, since i need to import scipy at the beginning and there is no way to check if it is supported or not

File: kratos/tests/test_sparse_matrices.py (hunk 8 above)
Comment (RiccardoRossi) [reply kratos/tests/test_sparse_matrices.py:108]: ok

File: kratos/tests/test_sparse_matrices.py (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/tests/test_sparse_matrices.py:89]: the point is that i am not very clear wheter scipy will be imported or not (possibly not, since it is a large dependency), however it is nice to have the code there as a sort of doc

File: kratos/python/add_sparse_matrices_to_python.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/python/add_sparse_matrices_to_python.h:14]: ok

File: kratos/python/add_sparse_matrices_to_python.cpp (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/python/add_sparse_matrices_to_python.cpp:201]: i would like to add the ones that are commented, however they need move semantics for ublas vectors (see #9541 )

File: kratos/tests/test_sparse_matrices.py (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/tests/test_sparse_matrices.py:89]: ok never mind, i found a solution

**Review by loumalouomega — APPROVED** [review]
Unblock

File: kratos/python/add_matrix_to_python.cpp [hunk 9]
```diff
@@ -191,6 +191,11 @@ namespace Python
 
           return matrix;
         }));
+
+
+        py::implicitly_convertible<py::buffer, DenseMatrix<double>>();
```
Comment (philbucher) [inline kratos/python/add_matrix_to_python.cpp:196]: I am guessing this could cause the tests in the LinearSolversApp to fail
The rest seems unrelated

Comment (RiccardoRossi) [reply kratos/python/add_matrix_to_python.cpp:196]: this one is simply telling that you can pass a numpy matrix to a function that expects a DenseMatrix

it should work silently...(and it has nothing to do with the linear solvers app...)

Comment (philbucher) [reply kratos/python/add_matrix_to_python.cpp:196]: true, now I saw it also failing in other PRs
